### ISSUE TITLE
jpeg image can't open

### ISSUE THREAD

**Issue by w136111526** [t0rakka/mango#14]
When I use Memory and Bitmap to save images, saving PNG pictures is normal, saving JPG pictures can not be opened. How can I modify my code?
my code as following:
	FILE* pFile = fopen("01.jpg", "rb");
	fseek(pFile, 0, SEEK_END);
	int iLen = ftell(pFile);
	vImg.resize(iLen);
	rewind(pFile);
	fread(vImg.data(), 1, vImg.size(), pFile);
	fclose(pFile);
	Memory mem(vImg.data(), vImg.size());
	auto surf = Bitmap(mem, ".jpg", FORMAT_R8G8B8);
	surf.save("E:\\test.jpg");	//can't open
	surf.save("E:\\test.png");	// normal

**Comment by t0rakka** [comment]
Added optimized path for encoding from 24 bit RGB but forgot to add support for it when writing the markers; oops! Fixed now!

**Comment by t0rakka** [comment]
That used to be possible when we still had 3 different "classes" of color images (rgb, luminance and palette/indexed). The color classification was deprecated 6+ years ago since the code paths were not used; the only useful target format (for realtime gfx / asset loading) was rgb. The Luminance -> RGB happens to work with the blitters because it is a trivial channel replication and uses existing code as-is. The RGB to luminance requires a dot product between the source color and luminance-weight and is not supported anymore. The idea is that we can load nearly every data imaginable and express it's memory layout and format precisely in a format that is convertible to RGB(A). If some format cannot be expressed as such then we pre-convert when we are loading, for example, PALETTE formats (indexed png, gif, etc) are converted to 32 bit BGRA when loading (the alpha is needed for opacity information).

**Comment by t0rakka** [comment]
Luckily the RGB -> Luminance conversion is trivial in itself; it gets more tricky when all different combinations has to be supported (float16 color, float32 color, 8, 12, 16 bit UNORM color, etc..) from any possible RGB permutation that is possible. The number of permutations was the primary reason for deprecation; the code did exist but it still is a bit heavy since we cannot JIT-generate it anymore. Our old light-weight system for that was x86-only and I did play around LLVM, which is actually fantastic for that, for generating the converters but it adds 20+ MB to the library size and takes "forever" to compile.

**Comment by t0rakka** [comment]
The current blitter is a bit work-in-process kinds of a thing. I have worked on a improved blitting system which is 100% SIMD friendly. I benchmarked the prototype and it is roughly twice the speed of the current code but there is a problem I haven't overcome yet: just for UNORM color we support 8, 16, 24, 32 bit source and target samples which means every loop needs 16 variations. Such code is easy to generate with templates but I don't want to explode the binary size so was playing around with separating samplers with converters, basically 3 function calls per sample. This drops the performance into one third or so. We probably could do a compromise and process in batches of 16 pixels or so and have a scratch buffer small enough to fit into L1 cache. If that works out well, the separation would be feasible from performance point of view and we could support a lot more blitters with less code and more importantly with less bloat.

**Comment by t0rakka** [comment]
That means the indexed and luminance support might be back.. eventually! :) Meanwhile, you could use the old good old NTSC perceived-luminance formula for the conversion:

L = 0.299 * R + 0.587 * G + 0.114 * B

Floating-point conversions might be expensive for your compiler (on x86-64 the default is SSE2 so the conversion instructions can have explicit rounding mode. which makes the conversion a lot cheaper single-instruction job). The reason for slow conversion with x87 (fpu stack) is that the current rounding mode must be preserved, so it must be stored, set, value converted and old mode has to be restored - that is a lot of code to do the simplest of things. A lot of old code emit "fistp" instruction directly to avoid the hassle but let's not go there. This is a completely non-issue if you do the conversion in fixed-point:

L = 0.299 * R + 0.587 * G + 0.114 * B

* 8192
// 8 bit RGB to 8 bit Luminance
L = ((R * 2449 + G * 4809 + B * 934) + 4096) >> 13;

Something like that should be a bit quicker. Simply reserve same amount of pixels in target format and do the above tranformation to every pixel. It is recommended that your innerloop is one scanline of the image because access to pixels is sequential. Don't use any putpixel()/getpixel() abstraction even if it is tempting (we have the address method in the surface but it is not means for those function, it is more useful for block operations like the scanline method). 

To keep things easier you should setup the source data to be in a known format; I see you already use the load-into-specific format constructor, which is nice. You can also convert later with the blit() method.

**Comment by w136111526** [comment]
look at mango's code shows that there is no way to write a grayscale image directly to a local file. Do you mean that I need to load the image into memory , and convert to a grayscale image using the blit() method?

**Comment by w136111526** [comment]
In addition, I see that you can quickly process zip files in your code, such as:
Bitmap bitmap(`` hello.zip / image. jpg ", Format(32, Format:: UNORM, Format::: BGRA, 8, 8, 8, 8); Can I read binary resources in visual studio C++,
 for example：
	HRSRC hResource = ::FindResource(NULL, MAKEINTRESOURCE(IDR_RES), _T("ZIPRES"));//IDR_RES is ZIP Resource file
	if (hResource != NULL) {
		DWORD dwSize = 0;
		HGLOBAL hGlobal = ::LoadResource(NULL, hResource);
		if (hGlobal != NULL) {
			dwSize = ::SizeofResource(NULL, hResource);
			if (dwSize > 0) 
			{
				//can I use mango::zip
			}
		}
		::FreeResource(hResource);
	}

**Comment by t0rakka** [comment]
btw. I probably should re-introduce the indexed and luminance conversions, even if only for 256-color-to-palette and 8 bit luminance, for convenience. Might hack up something together in the weekend. No promises.

**Comment by t0rakka** [comment]
A bit more about the grayscale image saving: zpng, png and jpeg write luminance format images, as long as you have them in that format in memory. JPEG only does 8 bit luminance and PNG/ZPNG do 8 and 16 bit.

**Comment by t0rakka** [comment]
OK; I made a prototype.. the gateway drug was the combination of Path and Mapper classes. I exposed the ZIP container AbstractMapper creation with getMemoryMapper() which is a new member function.

Then the Path interface uses the getMemoryMapper() to create the mapper on demands, namely, in the new constructor that takes  Memory and file-extension as parameter. The File is built on top of that; it simply creates a new temporary Path object in the new File constructor that also takes Memory and file-extension.
The biggest action happening in the getMemoryMapper, which is the only new function that does anything substantial. Of course, the mapping is valid only for the duration the Windows Resource is mapped, that goes without saying but just mentioning it so that it is more obvious. :)

There is ONE caveat, the container extension has to be expressed as-if it were part of a path: ".zip/", I could relax this rule so that just "zip" , "rar", etc. would work also but in the initial rushed-release that's the rule. 

Here is a sequence of code I used for verifying that the code does what it is supposed to. I developed this on Linux machine so the WIN32 magic is simulated but should work; the good thing is we don't need any Windows specific extensions for this to just work.

I'll push the code soon.

```
void memzip_test()
{
    // pretend this is a locked windows resource file..
    File file("xxx.zip");
    Memory mem = file;

    // create a "mapping" into the memory block in "zip" container format..
    Path path(mem, ".zip/");

    // list all the files in the container..
    for (auto x : path)
    {
        printf("%s (%d bytes) \n", x.name.c_str(), int(x.size));
    }

    // create a file view into the container...
    File f(mem, ".zip/", "test1.jpg");

    // ... because we don't have method in Bitmap class to do it otherwise: it needs
    // either a File or Memory view into the memory mapped image data, so this is how we get it done
    Bitmap bitmap(f, "jpg");
    bitmap.save("out.jpg");
}

```

**Comment by t0rakka** [comment]
If I figure out a nicer interface, I reserve the right for any changes for better. ;----)

**Comment by t0rakka** [comment]
The new code is pushed, check it out! 

p.s. The ridiculous ".zip/" rule has been relaxed, just use ".zip" or "zip", of course the fully decorated name also works. While working on the code I saw some spots that could be upgraded to better standards but that'll come later. Let me know  how this worked out with the Windows resources. :)

**Comment by t0rakka** [comment]
Latest example code for the "memory" as Path source:

```
Memory mem; // you provide this.. WIN32..
Path path(mem, "zip"); // it's a "zip" file.. 

// after that you use the Path object normally..
File file(path, "foo/bar.jpg"); // open a file in the zip..
Path path2(path, "foo/"); // sub-pathing works like before..
Bitmap picture(path, "flower.png");
// ... etc ...

```
TL;DR, the Path object is a Path object like others. The difference is that the root is the memory mapped container. It should just work.

**Comment by t0rakka** [comment]
Please do let me know if it works so that can close this. :)

**Comment by w136111526** [comment]
Okay, I 'll test your program later, I 'll try to use the uzip library instead of mango as a resource file decompression, and zpng is a really good library, and I'm not familiar with it yet.

**Comment by w136111526** [comment]
My job is to do image processing. I usually process uint8 data, which is faster than converting to float and INT。
Also, recently, after studying your mango library and trying to replace my SSE code with it, I found that there was something wrong with the SIMD part.
For example, when Vector processes an 8bit integer, the _ mm_add_pi8() function is used directly on the 8-bit register. Is there a data overflow phenomenon? do (240 +102) -150, can I get the correct result? My general practice is as follows. Do I must have to do this?
__m128i Zero = _mm_setzero_si128();
	unsigned char  udata[32] = { 1, 2, 3, 4, 5, 6, 8, 9, 10, 14, 121, 150, 204, 122, 16,  43, 24,  22,  184, 294, 22, 4, 11, 29, 41, 24,4,2,22,12,97 };
	__m128i FirstP0 = _mm_unpacklo_epi8(_mm_loadl_epi64((__m128i *)(udata + 0)), Zero);
	__m128i FirstP1 = _mm_unpacklo_epi8(_mm_loadl_epi64((__m128i *)(udata + 8)), Zero);
	__m128i FirstP2 = _mm_unpacklo_epi8(_mm_loadl_epi64((__m128i *)(udata + 32)), Zero);
	__m128i add16 = _mm_sub_epi16(FirstP2, _mm_add_epi16(FirstP0, FirstP1));
	__m128i pk8 = _mm_packus_epi16(add16, Zero);
	_mm_store_si128((__m128i*)dsum1, pk8);

**Comment by t0rakka** [comment]
The code looks fine; you expand the range and clamp to [0, 255] at packus_epi16. If you need more range you have to expand, right? =)

The conversion costs a little but the biggest win in sticking with 16 bits is that you double the number of elements processed per instruction vs. using full 32 bit elements. Looks fine..
